File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(0);
        ListNode dummy = res;
        ListNode temp1 = l1;
        ListNode temp2 = l2;
        int carry = 0;
        int sum = 0;
        while (temp1 != null || temp2 != null){
            int val1 = temp1 != null ? temp1.val : 0;
            int val2 = temp2 != null ? temp2.val : 0;
            int x = val1 + val2 + carry;
            carry = x / 10;
            sum = x % 10;
            dummy.next = new ListNode(sum);
            dummy = dummy.next;
            if (temp1 != null) temp1 = temp1.next;
            if (temp2 != null) temp2 = temp2.next;
        }
        if (carry != 0){
            dummy.next =  new ListNode(carry);
        }
        
        return res.next;
    }
}

File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;

        ListNode* head = l1;
        ListNode* prev = nullptr;

        while (l1 && l2) {
            int sum = l1->val + l2->val + carry;
            carry = sum / 10;
            l1->val = sum % 10;

            prev = l1;
            l1 = l1->next;
            l2 = l2->next;
        }

        if (!l1) {
            prev->next = l2;
            l1 = l2;
        }

        while (l1) {
            int sum = l1->val + carry;
            carry = sum / 10;
            l1->val = sum % 10;

            prev = l1;
            l1 = l1->next;
        }

        if (carry) {
            prev->next = new ListNode(carry);
        }

        return head;
    }
};